Generate e2e tests pipelines

thbkrkr opened the pull request:
This adds a go pipeline generator for running e2e-tests in Buildkite.

This allows to:
- centralize the definition of all nightly e2e tests (see .buildkite/e2e/nightly-main-matrix.yaml)
- support triggering any e2e tests from a PR comment
- avoid duplications in the buildkite pipeline
- remove the dependency on `.ci/setenconfig`

The generator can take `stdin` or command-line flags as arguments to define the e2e tests to run.
It can display the result as a buildkite pipeline or an environment file to reproduce a single run locally.

Mixing the two, it gives 3 modes:
- nightly-main/merge-main/pr-commit e2e tests: `cat matrix.yml | pipeline-gen | buildkite-agent pipeline upload`
- pr-comment e2e-tests: `pipeline-gen "$GITHUB_PR_TRIGGER_COMMENT_ARGS" | buildkite-agent pipeline upload`
- dev: `pipeline-gen -o envfile -f p=gke,k=1.23,s=8.6.0,t=TestSmoke | tee ../../../.env`

Tests suite runs can be organized in groups. Each group has an optional label, fixed environment variables shared by all the combinations in the group and mixed environment variables that are a list of environment variables for each combination. By just specifying environment variables, it can be verbose but this is very flexible. The pipeline generator doesn't validate uniqueness of tests combinations or verify the environment variables. A variable defined in the `fixed` field can be overridden in the `mixed` field. The only required variable for each combination is `E2E_PROVIDER`.

```yaml
- label: stack
  fixed:
    E2E_PROVIDER: gke
  mixed:
    - E2E_STACK_VERSION: "8.6.1"
    - E2E_STACK_VERSION: "8.7.0-SNAPSHOT"
      BUILD_LICENSE_PUBKEY: dev

- label: kind
  fixed:
    E2E_PROVIDER: kind
    TESTS_MATCH: TestSmoke
  mixed:
    - DEPLOYER_KIND_NODE_IMAGE: kindest/node:v1.21.12@sha256:f316b33dd88f8196379f38feb80545ef3ed44d9197dca1bfd48bcb1583210207
    - DEPLOYER_KIND_NODE_IMAGE: kindest/node:v1.22.9@sha256:8135260b959dfe320206eb36b3aeda9cffcb262f4b44cda6b33f7bb73f453105
    - DEPLOYER_KIND_NODE_IMAGE: kindest/node:v1.22.9@sha256:8135260b959dfe320206eb36b3aeda9cffcb262f4b44cda6b33f7bb73f453105
      DEPLOYER_KIND_IP_FAMILY: ipv6
```

I've added 'p,k,s,t' shortcuts for the most used variables to make PR comment more usable.

```
buildkite test this -f p=gke,k=1.23,s=8.6.0,t=TestSmoke
```

This is the final piece to stop relying on `setenvconfig`. Its logic is now separated into 3 parts:
- build vars used by steps in `.buildkite/build/setenv.sh` (introduced in #6377)
- tests runner vars are directly managed in the pipeline generator
- deployer vars are in `.buildkite/scripts/test/set-deployer-config.sh`

The only variable shared by the tests runner and the deployer is `CLUSTER_NAME` and is now generated by the pipeline generator instead of being set by us.

There are two specificities when not in CI, `make (operator|e2e)-image` is used instead of Buildkite metadata to set the images to use, and test license secrets are written to disk instead of being held in memory.

The generation and the upload of the JUnit XML report for each e2e tests run is enabled.

Some commands to test:
```sh
cat ../nightly-main-matrix.yaml | go run main.go | tee pipeline-nightly.yml

echo '- label: "kind/TestSmoke"
  fixed:
    E2E_PROVIDER: kind
    TESTS_MATCH: TestSmoke
' | go run main.go | tee pipeline-pr.yml

go run main.go -o envfile -f p=kind,t=TestSmoke
```

thbkrkr commented:
buildkite test this

```yaml
- label: stack
  fixed:
    E2E_PROVIDER: kind
  mixed:
    - E2E_STACK_VERSION: "6.8.23"
    - E2E_STACK_VERSION: "7.13.4"
 ```

thbkrkr commented:
buildkite test this -f p=gke,k=1.23,s=8.6.0,t=TestSmoke

thbkrkr commented:
buildkite test this -f p=gke,k=1.23,s=8.6.0,t=TestSmoke

thbkrkr commented:
buildkite test this -f p=gke,s=8.6.0,t=TestSmoke -m k=1.22,k=1.23

thbkrkr commented:
buildkite test this -f p=gke,s=8.6.0,t=TestSmoke -m k=1.22,k=1.23

thbkrkr commented:
buildkite test this -f p=gke,s=8.6.0,t=TestSmoke -m k=1.22,k=1.23,k=1.24

thbkrkr commented:
buildkite test this -f t=TestSmoke -m p=gke:s=8.7.0-SNAPSHOT:BUILD_LICENSE_PUBKEY=dev,p=eks,p=aks,p=tanzu,p=ocp:DEPLOYER_CLIENT_VERSION=4.10.34

pebrc commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+		}}
+
+	} else {
+		in, err := io.ReadAll(os.Stdin)
```
So when reading from a pipe all `-f` `-m` flags are ignored. Should we let the user know if case they specified some that they are without effect?

pebrc commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+		}
+	}
+
+	if output == "envfile" {
```
Should we validate that the chosen `output` is a supported value. If I put anything but `envfile` I get buildkite which is fine but then maybe it should just be a `-e|--envfile` flag?

thbkrkr replied:
I like the `-e|--envfile` that avoids the validation.

pebrc commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+	mixedsStr := strings.Split(str, ",")
+	for _, mixedStr := range mixedsStr {
+		mixedEnv := Env{}
+		envStr := strings.Split(mixedStr, ":")
```
This means that `mixed` specified as flags cannot contain docker images with `:` to separate image from tag?

thbkrkr replied:
Very good point. At first, I only supported one single env var per combination in `-f` but eventually found it a shame not support multiple variables. Then I wanted to avoid `;`, `&` or `|` that needs double quotes because they are special shell characters.

How about `^`?

`pipeline-gen -f p=gke -m k=1.22^s=8.5.0,k=1.23^s=8.6.0`

pebrc replied:
I think that would be OK and could be read as a logical conjunction operator which kind of fits the purpose.

pebrc commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+
+	operatorImageSuffix := os.Getenv(EnvVarBuildLicensePubkey)
+	if operatorImageSuffix != "" {
+		operatorImage = fmt.Sprintf("%s-%s", operatorImage, operatorImageSuffix)
```
This needs at least a comment to explain why we are appending the value of a variable that is judging by its name about license to the operator image. 

I am assuming the values are either `dev` or `` but still I am confused.

thbkrkr replied:
It's not useful today but the values could be anything.

I can add a comment but I admit I am not very satisfied with my refactoring of the operator image build to test snapshots. I think the main issue is that I use the variable `BUILD_LICENSE_PUBKEY` for 2 things: to discover the pub key to use but also implicitly to suffix the image, which is confusing I agree.

```js
pipeline-build
  if nightly-build
    additional build-operator-image with BUILD_LICENSE_PUBKEY
      -> use BUILD_LICENSE_PUBKEY to get a special pubkey embedded in the operator (cmd/license-initializer)
      -> push image suffixed by -$BUILD_LICENSE_PUBKEY

pipeline-e2e
  pipeline-gen
    if BUILD_LICENSE_PUBKEY
      -> use image suffixed by -$BUILD_LICENSE_PUBKEY
```

We could add a more explicit OPERATOR_IMAGE_SUFFIX which will be a repeat of BUILD_LICENSE_PUBKEY during build and will be used for testing instead of BUILD_LICENSE_PUBKEY?

```js
pipeline-build
  if nightly-build
    additional build-operator-image with BUILD_LICENSE_PUBKEY and OPERATOR_IMAGE_SUFFIX
      -> use BUILD_LICENSE_PUBKEY to get a special pubkey embedded in the operator (cmd/license-initializer)
      -> push image suffixed by -$OPERATOR_IMAGE_SUFFIX

pipeline-e2e
  pipeline-gen
    if OPERATOR_IMAGE_SUFFIX
      -> use image suffixed by -$OPERATOR_IMAGE_SUFFIX
```

This makes me realize that if we want to test snapshot stack version from a PR, the `-dev`operator image is currently not built.

pebrc replied:
Do you plan to address this in this PR or a subsequent one? Otherwise I think this PR looks 👍 to me.

thbkrkr replied:
I opened https://github.com/elastic/cloud-on-k8s/issues/6567 to address this in another PR.

pebrc reviewed (commented):
This looks really good! I have not actually tested whether the output is working but I guess you already did that in your PR through comments that triggered builds.

thbkrkr commented on .buildkite/scripts/test/set-deployer-config.sh:
```diff
@@ -0,0 +1,69 @@
+w()  { echo "$@" >> "$ROOT/deployer-config.yml"; }
+
+write_deployer_config() { 
+    :> "$ROOT/deployer-config.yml"
```
Absolutely. I had thoughts about this but it didn't seem trivial to me so I stuck to the script. But I think it's really something good to do to remove this intermediate layer written in bash.
I just opened a issue:
- https://github.com/elastic/cloud-on-k8s/issues/6561

pebrc replied:
Yes I think that is a good follow up task. Let's keep the bash script for now.

thbkrkr commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+	flag.StringVar(&output, "o", "buildkite-pipeline", "Type of output: buildkite-pipeline or envfile")
+	flag.Parse()
+
+	rand.Seed(time.Now().UTC().UnixNano())
```
This is used to generate a short identifier for the cluster name to make it unique. I see it as an additional protection if we make mistakes declaring duplicate tests combinations or if the generator has a bug due to a particular mishandled use case.

```sh
> cat ../nightly-main-matrix.yaml| go run main.go| grep -o "CLUSTER_NAME:.*" | grep stack | sort | uniq
CLUSTER_NAME: "eck-e2e-stack-6-8-23-vgoj-0"
CLUSTER_NAME: "eck-e2e-stack-7-13-4-nmuo-0"
CLUSTER_NAME: "eck-e2e-stack-7-14-2-pqxv-0"
CLUSTER_NAME: "eck-e2e-stack-7-15-2-aovq-0"
CLUSTER_NAME: "eck-e2e-stack-7-16-3-qgux-0"
CLUSTER_NAME: "eck-e2e-stack-7-17-8-aqbi-0"
CLUSTER_NAME: "eck-e2e-stack-8-0-1-cmkp-0"
CLUSTER_NAME: "eck-e2e-stack-8-1-3-klks-0"
CLUSTER_NAME: "eck-e2e-stack-8-2-3-tmtk-0"
CLUSTER_NAME: "eck-e2e-stack-8-3-3-segs-0"
CLUSTER_NAME: "eck-e2e-stack-8-4-3-xard-0"
CLUSTER_NAME: "eck-e2e-stack-8-7-0-snap-ipaj-0"
```

thbkrkr replied:
You are right. I just didn't know. I was about to replace it with `randomGen = rand.New(rand.NewSource(time.Now().UTC().UnixNano()))` but the default random seed is fine here.

thbkrkr commented on .buildkite/e2e/pipeline-gen/main.go:
```diff
@@ -0,0 +1,390 @@
+		in, err := io.ReadAll(os.Stdin)
+		handlErr("Failed to read stdin", err)
+		if len(in) == 0 {
+			handlErr("Failed to read stdin", errors.New("nothing on /dev/stdin"))
```
I will change because I recognize that it may look like a typo, but it was intentional at first.

thbkrkr commented:
buildkite test this -f t=TestSmoke -m p=aks,p=eks,p=gke,p=kind,p=tanzu,p=ocp^DEPLOYER_CLIENT_VERSION=4.10.34

thbkrkr commented:
buildkite test this -f t=TestSmoke -m p=aks,p=eks,p=gke,p=kind,p=tanzu,p=ocp

thbkrkr commented:
buildkite test this -f t=TestSmoke -m p=aks,p=eks,p=gke,p=kind,p=tanzu,p=ocp

thbkrkr commented:
I implemented the split between the cluster creation and the e2e-tests execution. It's important for ocp and tanzu where the cluster creation takes longer (see https://github.com/elastic/cloud-on-k8s/pull/6525#issuecomment-1477458576). I implemented a list of providers that don't support remote kubeconfig to exclude kind that is not ready for that.

My first attempt with tanzu don't work: `Error: unknown command "cluster" for "tanzu"`.

https://buildkite.com/elastic/cloud-on-k8s-operator/builds/2506#01870b1b-1722-4321-9335-5170b40120ce

```
./hack/deployer/deployer get credentials --plans-file hack/deployer/config/plans.yml --config-file deployer-config.yml
2023/03/22 21:22:34 Logging in to Azure
2023/03/22 21:23:06 Logging in to container registry
2023/03/22 21:23:12 Installing Azure storage-preview extension
WARNING: The installed extension 'storage-preview' is in preview.
2023/03/22 21:23:25 Defaulting WorkDir: /root/eck-e2e-tanzu-fuvk-25062862031132
2023/03/22 21:23:25 Using installer state dir: /root/eck-e2e-tanzu-fuvk-25062862031132
2023/03/22 21:23:29 Reusing existing storage container to persist installer state
2023/03/22 21:23:29 Restoring installer state from storage container if any
2023/03/22 21:23:33 Installing Tanzu CLI plugins
Unable to find image 'cloudonk8s.azurecr.io/tanzu-ci:1.5.4' locally
1.5.4: Pulling from tanzu-ci
..
Digest: sha256:2e84efb55f2895d2deb4d87cc617efba094f68e4c24bbafc09d071b4ea17950c
Status: Downloaded newer image for cloudonk8s.azurecr.io/tanzu-ci:1.5.4
Checking for required plugins...
Installing plugin 'login:v0.11.6'
...
Successfully installed all required plugins
✔  successfully initialized CLI
Error: unknown command "cluster" for "tanzu"
Run 'tanzu --help' for usage.
 
✖  unknown command "cluster" for "tanzu"
Error: exit status 1
exit status 1
make: *** [Makefile:586: set-kubeconfig] Error 1
```

I moved tanzu in the list of providers that don't support remove kubeconfig, which is not true but just for testing and now I got:

```
Error: unable to set up management cluster: unable to wait for cluster and get the cluster kubeconfig: error waiting for cluster to be provisioned (this may take a few minutes): cluster creation failed, reason:'VMProvisionFailed @ Machine/tkg-mgmt-azure-20230322222434-control-plane-5cm5s', message:'1 of 2 completed'
```

I suspect a provisioning issue due to tanzu clusters created by this PR not being removed due to a bug I introduced that broke cleanup.

thbkrkr commented:
buildkite test this -f t=TestSmoke -m p=tanzu

thbkrkr commented:
Tanzu's latest creation succeeded, confirming it was a provisioning issue. I'm debugging why the split between creating the cluster and running e2e-tests doesn't work for this provider.

thbkrkr commented:
We discussed this offline but for visibility. 

> After the step that builds the tanzu cluster we should have a management cluster and therefore the commands should be available?

The management cluster exists but when the cli is used in a new box, the management cluster is no longer configured. It may be a problem with saving the install state.

thbkrkr commented:
buildkite test this -f t=TestSmoke,p=tanzu,TEST_LICENSE=, MONITORING_SECRETS =

This is an attempt to see if we can run TestSmoke in tanzu without any test license secrets.

thbkrkr commented:
buildkite test this -f t=TestSmoke,p=tanzu,TEST_LICENSE=,MONITORING_SECRETS=

thbkrkr commented:
buildkite test this -f t=TestSmoke,p=tanzu,TEST_LICENSE=,MONITORING_SECRETS=,GO_TAGS=
